Dependence graph: ensure grapht representation is consistent with domain

smowton opened the pull request:
Previously the grapht representation was incrementally populated at the same time as the domain representation was computed, but at least one case (control dependencies crossing function callsites) was missing, leading to inconsistency (and so full-slicer incorrectness). Patching it up and keeping the pattern invites future similar mistakes. Instead, this commit switches to generating the grapht representation in one go after the AI fixedpoint has been computed.

smowton committed:
Abstract interpreter: add finalize hook

At least dependence_grapht has good reason to do work at the end of the day--
namely to create an alternative data representation that is trickier to keep
up to date incrementally than to produce in one go at the end. It seems likely
other AIs will have use for a post-fixedpoint hook.

smowton committed:
Dependence graph: ensure grapht representation is consistent with domain

Previously the grapht representation was incrementally populated at the same
time as the domain representation was computed, but at least one case (control
dependencies crossing function callsites) was missing, leading to inconsistency,
and doing it this way invites future similar mistakes. Instead, this commit switches
to generating the grapht representation in one go after the AI fixedpoint has been
computed.

tautschnig commented:
This is a great find, but could you please add a test case? Your description hints at the existing of such?

smowton committed:
Add unit test checking dependence graph consistency

This ensures that the dependence graph's internal representation
has been accurately transcribed to its grapht external interface,
and exposes the particular case of control dependencies being
preserved across callsites.

smowton commented:
@tautschnig done, added a unit test checking the particular case of function calls with control dependencies

martin-cs commented:
Thank you @smowton ; most helpful.  This might help @lucasccordeiro with slicer bugs...

lucasccordeiro commented:
Thanks, @martin-cs.

There has been some progress on the security analyzer team in using the full slicer. Some weeks ago, @marek-trtik and I have investigated one particular issue, which ended up in a fix in the dependence graph. @marek-trtik: Was your PR merged into the develop branch? 

BTW, I met Jens Krinke at ASE'17. We're currently planning a talk at Diffblue abour `dynamic program slicing` by Jens.

marek-trtik commented:
@lucasccordeiro : No, the PR was only merged into CBMC sub-tree of the `security-scanner` and it is still only there.
